Usage cost implementation

satyamakgec opened the pull request.

SatyamSB committed:
re addition of usage cost logic

SatyamSB committed:
add multi signature interface

SatyamSB committed:
improve test cases

SatyamSB committed:
add test for the usage fee

SatyamSB committed:
add multi sig functionality in STR

SatyamSB committed:
fix tests and interface

SatyamSB committed:
resolve conflicts

SatyamSB committed:
to fix coverage issue

SatyamSB committed:
add whitelabler functionality in setup cost

SatyamSB committed:
resolve conflicts

SatyamSB committed:
usage cost implementation

SatyamSB committed:
fix tests

SatyamSB committed:
remove unncessary test code

maxsam4 commented on contracts/modules/Module.sol:
```diff
@@ -81,12 +85,25 @@ contract Module is IModule, ModuleStorage, Pausable {
+     * @notice Deduct the usage cost
+     */
+    function _deductUsageFee() internal {
+        uint256 _usageCost = IModuleFactory(factory).usageCostInPoly();
```
This means that the factory owner can arbitrarily change usage cost. Front running attacks are also possible where the issuer initiates the transaction thinking that fee is X but module owner front runs that tx to increase the usage fee to Y.

maxsam4 replied:
The issuer can reduce the risk by properly using the budgeting in the security token but I think we can come up with a better solution.

satyamakgec replied:
lockup period is used to avoid front running issues. Factory owner can change the value of usage cost in 24 hours only because of this front running chances are very low.

maxsam4 replied:
In its current implementation, the lock period is failing to do that. It's just limiting the module owner to do one front-run attack per 24 hours. Ideally, Even one attack shouldn't be possible.

If we implement something like what I mentioned in the other review comment, both this and that review comment will be solved.

satyamakgec replied:
hmm, Both cases are probabilistic but anyway I don't have any problem to implement.

maxsam4 commented on contracts/modules/ModuleFactory.sol:
```diff
@@ -28,6 +31,9 @@ contract ModuleFactory is IModuleFactory, Ownable {
     uint256 public setupCost;
+    uint256 public usageCost;
+    uint256 public lastTimeUsageCostChange;
+    uint256 internal constant LOCKPERIOD = 24 hours;
```
I am not sure how/why LOCKPERIOD is useful in its current form.

Perhaps we should do something like
1) user can change fee at any time.
2) new fee will become active only after lockperiod.

adamdossa replied:
Agreed - should be a two step propose:
  - factory owner proposes a usage fee change
  - factory owner can commit / finalise the change 24 hours after it has been proposed

Exception is when the module is first created of course.

adamdossa replied:
NB - if we did this for setup & usage costs, we could get rid of the logic in ST where the user has to pass an expected setup amount as well - not sure if this is worth fiddling with at this point.

VictorVicente committed:
CLI - Added usageCost and budget support

SatyamSB committed:
initial variables naming

adamdossa commented on contracts/modules/Module.sol:
```diff
@@ -81,12 +85,25 @@ contract Module is IModule, ModuleStorage, Pausable {
+        uint256 _usageCost = IModuleFactory(factory).usageCostInPoly();
+        if (_usageCost > 0) {
+            address factoryOwner = IModuleFactory(factory).owner();
+            require(factoryOwner != address(0), "Invalid owner");
```
I guess this means we can break modules (that have usage fees) by setting the factoryOwner to 0x0? I wonder whether we should just add this constraint to `if (_usageCost > 0)` instead so that if someone sets the factory owner to 0x0 they lose fees, but don't break the module?

satyamakgec replied:
That makes sense

adamdossa commented on contracts/modules/Experimental/TransferManager/SignedTransferManagerFactory.sol:
```diff
@@ -10,13 +10,18 @@ contract SignedTransferManagerFactory is ModuleFactory {
         bool _isCostInPoly
     )
-        public ModuleFactory(_setupCost, _polymathRegistry, _isCostInPoly)
+        public ModuleFactory(_setupCost, _usageCost, _polymathRegistry, _isCostInPoly)
```
General comment is that with this design we can only have one usage fee per module - so any actions in that module must cost the same amount of POLY. If this is OK w/ product that's fine, but not 100% sure if that's the case or not.

satyamakgec replied:
I do have a similar doubt in my mind, but if you see all our modules they are pretty focussed on one major activity and revolves around that. Still, it is a good question to ask to the business team.

adamdossa commented on contracts/interfaces/IModuleFactory.sol:
```diff
@@ -115,4 +133,11 @@ interface IModuleFactory {
      */
     function changeTitle(string calldata _title) external;
 
+    function polymathRegistry() external returns (address polymathRegistryAddress);
```
Where is this used? Quick search doesn't show anything.

satyamakgec replied:
I used this in the earlier implementation but afterwards requirements get change. I keep this because it doesn't hurt anything and make it compliant with ModuleFactory contract.

adamdossa commented on contracts/modules/Checkpoint/Voting/APLCR/AdvancedPLCRVotingCheckpoint.sol:
```diff
@@ -485,6 +486,7 @@ contract AdvancedPLCRVotingCheckpoint is AdvancedPLCRVotingCheckpointStorage, Vo
             }
             choiceCount = temp;
         }
+        require(totalChoiceWeight <= weight);
```
Is this change due to moving some of the voting stuff to a library, or a separate bug you found?

satyamakgec replied:
Yeah it is related to the separate bug I found

VictorVicente committed:
CLI - Added token balance to ST summary.
CLI - Added remaining budget to module data.
CLI - Ask users if they want to transfer POLY when module budget change

SatyamSB committed:
implement coldperiod

maxsam4 commented on contracts/modules/ModuleFactory.sol:
```diff
@@ -94,10 +95,17 @@ contract ModuleFactory is IModuleFactory, Ownable {
+        if (now > lastTimeUsageCostChange.add(COLDPERIOD))
+            return usageFee;
+        else
+            return tempUsageFee;
```
This can be bypassed by changing the fee twice. We should modify the change usage cost function to only update the tempUsageFee if `now > lastTimeUsageCostChange.add(COLDPERIOD)`.

adamdossa replied:
Yeah - I tend to think of this as needing two operations from the factory owner:
1. set proposed new usage fee (emit UsageFeeProposed(proposedFee, currentFee))
2. confirm proposed usage fee (can only be called more than 24 hours after 1 above) (emit ChangeUsageCost(oldFee, newFee))

This means that if 2. never happens, the fee is never changed. I see this as something that would happen rarely, so not an issue that there needs to be two calls.

VictorVicente committed:
CLI - Don't ask budget if module has no usageCost

VictorVicente committed:
Merge branch 'usage-cost-implementation' of https://github.com/PolymathNetwork/polymath-core into usage-cost-implementation

VictorVicente committed:
CLI - Minor fixes

SatyamSB committed:
improve the process of changing usage cost

SatyamSB committed:
Merge branch 'usage-cost-implementation' of https://github.com/PolymathNetwork/polymath-core into usage-cost-implementation

SatyamSB committed:
addition of custom oracles

SatyamSB committed:
change the title of factory

maxsam4 commented on contracts/modules/ModuleFactory.sol:
```diff
@@ -88,24 +89,26 @@ contract ModuleFactory is IModuleFactory, Ownable {
-        _changeUsageCost(_usageCost);
+    function changeUsageCost() public onlyOwner {
+        require(now > usageCostProposedAt.add(COLDPERIOD), "Proposal is in unmatured state");
+        require(usageCost != proposedUsageCost);
```
This check seems unnecessary here. Perhaps, move it to _proposeUsageCost()?

maxsam4 reviewed (commented):
`changeCostAndType` allows owner to change `isCostInPoly` without any cooldown. Since 1 POLY != 1 USD, this can be used to arbitrarily increase (or decrease) usage cost. There should be a 24 hour CD for changing this as well.

SatyamSB committed:
add test suite for the STO

satyamakgec committed:
Merge branch 'dev-3.1.0' into custom-oracles

SatyamSB committed:
resolve conflicts in APLCR

SatyamSB committed:
proposed functionality in cost type

SatyamSB committed:
fix test in ctm

maxsam4 commented:
@satyamakgec I think some merge went wrong somewhere. I am seeing all kinds of changes here.

maxsam4 reviewed (commented):
The main changes look good. Please clean up the extra changes before merging.

pabloruiz55 commented:
#837 has been merged. Can the PR be closed? @satyamakgec

satyamakgec commented:
Yes @pabloruiz55 It can be closed
